Energy model-based tests, RFC v2

bjackman opened the pull request:
See the commit messages for the [first](https://github.com/ARM-software/lisa/commit/399cdfff3c112836408f03346832df52008779d7) and [fourth](https://github.com/ARM-software/lisa/commit/37918fade84d389a6df2183ac02892fe3fef898d) commits for more detail on what this PR aims to introduce. In short, this aims to allow automatically running tests for any given synthetic workload and platform. 

This looks like an enormous PR, and it is... but a lot of the added lines are `.ipynb` noise or boilerplate in the [provided](https://github.com/ARM-software/lisa/commit/399cdfff3c112836408f03346832df52008779d7#diff-6a86319666706547b6bb03baf38e4916) [platform](https://github.com/ARM-software/lisa/commit/399cdfff3c112836408f03346832df52008779d7#diff-ac2a5c8c78e008f3791e81447c911ca6) [descriptions](https://github.com/ARM-software/lisa/commit/399cdfff3c112836408f03346832df52008779d7#diff-b622e9d205b02984946d5cdb1cd96a96). I've also added pretty verbose docstrings which bumps up the line count a lot. 

I've siloed myself a bit on this so please highlight wherever I haven't made the high-level intention clear, It's important that someone can understand what this stuff is trying to do without reading 2k lines of inscrutable bjackman Python :smile: . Hopefully docstrings are a good starting point...

bjackman commented:
Another example use: https://github.com/bjackman/lisa/blob/cie-test/tests/eas/load_tracking.py#L247

bjackman commented:
Today I ran the "generic" tests I've added here and compared the energy estimation from `EnergyModel` with the energy measured with an ACME cape on Pixel and hwmon on Juno. 

The only numerical analysis I did was to compare the estimated and measured energy for the *entirety of each workload*. There's a strong correlation (>95% coefficient) there.

As for finer grained predictions (i.e. predicting& measuring the variation in energy consumption *during the course of the workload*), I plotted energy measurements over time (using ACME) and compared visually with plots of energy predictions over time. They look to match nicely but like I say this wasn't very scientific.

ionela-voinescu commented on ipynb/tests/Generic_EAS_Tests.ipynb:
```diff
@@ -0,0 +1,533 @@
+       "            /* TRAPPY_PUBLISH_REMOVE_STOP */\n",
+       "            \n",
+       "        req([\"require\", \"EventPlot\"], function() { /* TRAPPY_PUBLISH_REMOVE_LINE */\n",
+       "            EventPlot.generate('fig_a283b792c9da4efe9e175518453bf49b', '/nbextensions/', {\"lanes\": [{\"id\": 0, \"label\": \"CPU :0\"}, {\"id\": 1, \"label\": \"CPU :1\"}, {\"id\": 2, \"label\": \"CPU :2\"}, {\"id\": 3, \"label\": \"CPU :3\"}, {\"id\": 4, \"label\": \"CPU :4\"}, {\"id\": 5, \"label\": \"CPU :5\"}], \"colorMap\": null, \"keys\": [\"task_wmig0-1981\", \"task_wmig1-1982\", ...
```
It would be more clear if these would print a success message when the test is passed as there is no output from it otherwise. The same for test_task_placement.

bjackman replied:
True, will add this.

ionela-voinescu commented on ipynb/tests/Generic_EAS_Tests.ipynb:
```diff
@@ -0,0 +1,533 @@
+   },
+   "outputs": [],
+   "source": [
+    "utils = t.get_task_utils_df(experiment)"
```
While I understand why you decided not to print the data frames for this call and the ones below, I for one would like to have more information in these test notebooks as it gives me more confidence in the test. If you have some calls to some functions that do not return anything it might make people think that something is wrong and maybe the test is not doing what it's suppose to be doing. I think an option would be to print the traces and plots first and maybe have a debug section after that where the data frames are printed and the process is explained in a few sentences.

bjackman replied:
The functions do return something, it's just assigned to a variable, I wouldn't expect to see any output from running that code. But you're right that this is unclear.

How about if we print the docstring for the function before calling it, then you basically get a detailed description of what's going on.

ionela-voinescu replied:
Sorry! I did not make myself clear. I did not mean to say that the calls don't return anything but that you can't see anything being printed as a result of them being run. I suppose printing the docstrings could work but what I meant to say is that it would be nice to have a section where you can see more details on the test being run. Details on the workload, the data frames obtained, something that could represent the basis of a debugging process.

bjackman replied:
I don't intend for this to be for debugging the *tests*, after all they are absolutely flawless in every single way. Rather it's intended as a starting point for looking at what element of the kernel behaviour was judged to be faulty (since when you have a failure from a lisa-test run it's pretty much "computer says no").

Basically the only thing I mean for this to show is a plot of the workload task residency and a plot of the energy estimations, so you can see which part of the workload run was judged to be mis-scheduled. I don't imagine users diving python in order to do that, but rather just taking a look and seeing "oh yeah this 5% task ran on a big CPU at t=3.5" then going back to the kernel to figure out why that happened.

Nonetheless I'll add some more english to explain what is going on here.

ionela-voinescu commented on ipynb/tests/Generic_EAS_Tests.ipynb:
```diff
@@ -0,0 +1,533 @@
+    "collapsed": true
+   },
+   "source": [
+    "# Run Generic Automated EAS tests\n",
```
It would be good to have more information about what the tests report and how the evaluation is done. As you say below "These tests make estimation of the energy efficiency of task placements.. " + "by.. ". It would also be nice to make clear the distinction between the estimated observed power consumption (estimated based on the energy model provided and the observed task placement decisions made) and the expected power consumption (estimated based on the energy model provided and the expected optimal task placements). This can be done either here at the beginning or further in the notebook where the estimated and expected data frames are obtained.

bjackman replied:
True. I will try to clarify in the notebook.

bjackman commented on tests/eas/generic.py:
```diff
@@ -0,0 +1,433 @@
+            est_energy, exp_energy)
+        self.assertLess(est_energy, exp_energy * 1.2, msg=msg)
+
+class OneSmallTask(EnergyModelTest):
```
Why? In this case "generic" is supposed to mean "works on any platform".

ionela-voinescu replied:
I thought generic meant more that it would hold the base EnergyModelTest class, which then can be extended in specific test classes. I find it more clear to have the EnergyModelTest class in its own file but it's just a preference.

bjackman commented on tests/eas/generic.py:
```diff
@@ -0,0 +1,433 @@
+
+        msg = 'Estimated {} bogo-Joules to run workload, expected {}'.format(
+            est_energy, exp_energy)
+        self.assertLess(est_energy, exp_energy * 1.2, msg=msg)
```
Yep, will do.

bjackman commented on tests/eas/generic.py:
```diff
@@ -0,0 +1,433 @@
+                    cpu_utils[int(cpu)] += util
+            power = nrg_model.estimate_from_cpu_util(cpu_utils)
+            columns = power.keys()
+            return pd.Series([power[c] for c in columns], index=columns)
```
Yep. I think I'll remove the "power" column and sort the node columns by depth-first order.

bjackman commented on tests/eas/__init__.py:
```diff
@@ -0,0 +1,101 @@
+            cls.target.write_value(
+                "/proc/sys/kernel/sched_initial_task_util", 1024, verify=False)
+
+    def _do_test_first_cpu(self, experiment, tasks):
```
Yes. This function actually just being moved from acceptance.py so I'll leave it here for now. None of the new tests should use it.

bjackman commented on tests/eas/acceptance.py:
```diff
@@ -393,7 +408,10 @@ def test_all_tasks_finish_on_a_big_cpu(self, experiment, tasks):
+                          msg=msg)
 
 
 class WakeMigration(EasTest):
```
Will take a look.

bjackman commented:
OK, thanks for the review. I've responded to the comments to remind myself what changes I need to make, and I'll come back once I have time to make the code changes.

bjackman commented:
OK I've pushed a new version where I've tried to address your comments:

- Fixing the minor stuff
- I added some more text to the notebook to explain what's going on. Also some of the calls that you commented on actually weren't necessary so I removed them. Hopefully it's a bit more easy to understand (assuming people actually read my walls of text :grimacing:)
- Fixed the order of the columns in power DataFrames (For juno you get [(1), (2), (1, 2), (0), (3), (4), (5), (0, 3, 4, 5)] i.e. LITTLEs followed by LITTLE cluster then bigs followed by big cluster).
- Fixed the remaining `bl` references in `acceptance.py`.

ionela-voinescu commented on ipynb/tests/Generic_EAS_Tests.ipynb:
```diff
@@ -0,0 +1,585 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Maximum estimated system energy: 1725\n",
```
These are no longer printed (based on a commit merged a short while ago).

ionela-voinescu commented on ipynb/tests/Generic_EAS_Tests.ipynb:
```diff
@@ -0,0 +1,585 @@
+       "            /* TRAPPY_PUBLISH_REMOVE_STOP */\n",
+       "            \n",
+       "        req([\"require\", \"EventPlot\"], function() { /* TRAPPY_PUBLISH_REMOVE_LINE */\n",
+       "            EventPlot.generate('fig_dbb60bcbad814ee0ba22603291329244', '/nbextensions/', {\"lanes\": [{\"id\": 0, \"label\": \"CPU :0\"}, {\"id\": 1, \"label\": \"CPU :1\"}, {\"id\": 2, \"label\": \"CPU :2\"}, {\"id\": 3, \"label\": \"CPU :3\"}, {\"id\": 4, \"label\": \"CPU :4\"}, {\"id\": 5, \"label\": \"CPU :5\"}], \"colorMap\": null, \"keys\": [\"task_wmig0-3273\", \"rt-app-3272\", \"t ...
```
.. is higher than the **ideal energy plot**.. 
Although a reword of it might work: "Where the values for the observed energy are higher than the ideal energy values..". The same for the following sentence. It's strange to say "where the plot is higher". Maybe a better wording would be "where the plot line is higher.. ".

ionela-voinescu replied:
Awesome! :D

bjackman replied:
Good spot. Good point about "plot" vs "values" will reword this.

ionela-voinescu commented on ipynb/tests/Generic_EAS_Tests.ipynb:
```diff
@@ -0,0 +1,585 @@
+    "import logging\n",
+    "from conf import LisaLogging\n",
+    "LisaLogging.setup()\n",
+    "logging.getLogger().setLevel(logging.WARNING)\n",
```
Isn't this the same as saying: LisaLogging.setup(level=logging.WARNING)?

bjackman replied:
Yep, will change.

ionela-voinescu commented on tests/eas/generic.py:
```diff
@@ -0,0 +1,450 @@
+
+class EnergyModelWakeMigration(EnergyModelTest):
+    """
+    Test EAS for tasks alternating beetween 10% and 50%
```
"Test EAS for tasks alternating beetween 10% and 50% over 4 seconds". It would be good to specify the duration of the entire workload as for the time axis of the plot to be better understood. The same for the other classes. Some of them include this information but some of them do not.

bjackman commented:
Updated

bjackman commented on libs/utils/platforms/hikey_energy.py:
```diff
@@ -0,0 +1,68 @@
+from collections import OrderedDict
+
+cluster_active_states = OrderedDict([
+    (208000, ActiveState(capacity=178, power=16)),
```
Sure. I'm a little bit awkward about doing this because for some reason my PEP8 religion forbids it but TBH it is clearer so let's apply it here.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+    Just like EnergyModelNode except that ``active_states`` and ``idle_states``
+    aren't required.
+    """
+    def __init__(self, active_states=None, idle_states=None, *args, **kwargs):
```
Sure, it seems clearer.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+        Used throughout this module: A ``util_distrib`` is a list ``u`` where
+        ``u[N]`` is the sum of the frequency-invariant, capacity-invariant
+        utilization of tasks placed on CPU N. That is, the quantity represented
+        by a CPU runqueue's util_avg in the Linux kernel scheduler's
```
Good point, will add this note.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+
+          # All CPUs idle:
+          [0, 0, 0, 0] -> ["cluster-sleep-0", "cluster-sleep-0",
+                           "cluster-sleep-0", "cluster-sleep-0"]
```
Yep, good catch.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+
+        - 4 CPUs, with CPUs [0, 1] in PD0 and CPUs [2, 3] in PD1
+
+        - 4 idle states: "WFI", "cpu-sleep", "cluster-sleep-0" and "cluster-sleep-1"
```
Good point, will add this.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+          # Note that CPU 2 has no work but is assumed to never be able to enter
+          # any "cluster" state.
+          [0, 0, 0, 1] -> ["cluster-sleep-1", "cluster-sleep-1",
+                           "cpu-sleep","WFI"]
```
No:

> If a CPU has any tasks it is estimated that it may only enter its
        shallowest idle state in between task activations

And in line 346 all the CPUs have tasks so can only enter WFI

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+            if possible_freqs:
+                ideal_freqs[cpu] = min(possible_freqs)
+            else:
+                # CPU cannot provide required capacity, use max freq
```
No, I don't agree that this can never happen - the `util_distrib` is not capped (and IIUC `util_avg` in a trace can be greater than the current capacity).

derkling replied:
Right: actually the tracepoint reports the "raw" RQ's value:
https://android.googlesource.com/kernel/common/+/android-3.18/include/trace/events/sched.h#690

... thus yes, for example on task migrations, it can be bigger than the current capacity ;-)

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+            else:
+                # CPU cannot provide required capacity, use max freq
+                ideal_freqs[cpu] = max(node.active_states.keys())
+                overutilized = True
```
Hmm.. `overutilized` here isn't really the same thing as it is in EAS. I think adding an 80% limit would be a bit arbitrary as this is supposed to be  a theoretical estimation, whereas EAS's 80% is a limit to respect the fact that our util estimation is imperfect. I would prefer to add it as a parameter somewhere, but in a later iteration.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+        assert all(0.0 <= a <= 1.0 for a in cpu_active_time)
+
+        for node in self.root.iter_nodes():
+            if not node.active_states or not node.idle_states:
```
Yep, will add this.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+
+            cpus = tuple(node.cpus)
+            # For now we assume topology nodes with energy models do not overlap
+            # with frequency domains
```
Sure, I've added a check to `__init__` but I'm also going to leave this comment and assert here, if that sounds good to you.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+            # This works great for the synthetic periodic workloads we use in
+            # LISA (where all threads wake up at the same time) but is probably
+            # no good for real workloads.
+            active_time = max(cpu_active_time[c] for c in cpus)
```
Yeah, I actually have that on my other branch: https://github.com/bjackman/lisa/blob/energy-model-rfc-v2-evaluation/libs/utils/energy_model.py#L527 (ctrl+F util_aggregator) but decided to avoid the complexity here.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+            _idle_power = max(node.idle_states[idle_states[c]] for c in cpus)
+            idle_power = _idle_power * (1 - active_time)
+
+            if combine:
```
Hmm, I actually thought I had removed this parameter completely, it seemed useful at first but in the end I never ended up using it, would that be OK with you?

derkling replied:
OK, let's add it in future if we need to breakdown of the contributions...

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+                    {
+                        (0,)    : 10,
+                        (1,)    : 10,
+                        (1, 2)  : 5,
```
You are not wrong. Thanks again for all the checking of these examples.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+                  This represents CPUs 0 and 1 each using 10bW and their shared
+                  resources using 5bW for a total of 25bW.
+        """
+        if len(util_distrib) != len(self.cpus):
```
> What about `cpus_utilizations`?
Hmm, yeah I think that would be a sensible name.

> it seems that you want to support the concept of variable workloads using "a series of util_distrib" ... but from this check this seems not to be supported right now, sin't it?
The implication of using a series of `util_distrib`s was that you'd call this function for each of them; that's what is done in generic.py (see `get_power_df`). Perhaps I can reword to make that clearer.

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+                           single-phase periodic RT-App tasks is an example of a
+                           suitable workload for this model.
+        :returns: List of ``util_distrib`` s representing distribution of work
+                  under optimal task placments see
```
mm.. now I'm thinking about....

:balloon:  [**CI FOR SPELLING**](https://sphinxcontrib-spelling.readthedocs.io/en/latest/) :balloon:

bjackman commented on libs/utils/energy_model.py:
```diff
@@ -0,0 +1,572 @@
+            for task, cpu in placement.items():
+                util[cpu] += capacities[task]
+            util = tuple(util)
+
```
Nice, thanks!

bjackman commented on tests/lisa/test_energy_model.py:
```diff
@@ -0,0 +1,253 @@
+    ('WFI',              5),
+    ('cpu-sleep-0',      5),
+    ('cluster-sleep-0',  1),
+])
```
Sounds good.

I did also implement a [method to print various observations about the energy model](https://github.com/bjackman/lisa/blob/energy-model-rfc-v2-evaluation/libs/utils/energy_model.py#L793) but didn't include it here.

bjackman commented on tests/lisa/test_energy_model.py:
```diff
@@ -0,0 +1,253 @@
+                                                   [0, 1, 0, 0]])
+
+    def test_single_big(self):
+        placements = em.get_optimal_placements({'task0': 350})
```
Yeah, this was a deliberate decision at the beginning of writing it, because I thought if I expressed it in terms of the data I'd probably make the exact same mistakes as I'd made in the code under test. So I wanted to work them out by hand. 

However I agree that as the code grew it didn't start to look like such a smart idea. I'm going to keep it as it is for now, but add a comment. Probably if we need to extend these tests it will be faster to add new ones in terms of the data rather than go back and re-factor the hard coded stuff.

derkling commented:
On 23-Feb 07:20, Brendan Jackman wrote:
> bjackman commented on this pull request.
> 
> 
> 
> > +            # with frequency domains
> +            freq = freqs[cpus[0]]
> +            assert all(freqs[c] == freq for c in cpus[1:])
> +
> +            # The active time of a node is estimated as the max of the active
> +            # times of its children.
> +            # This works great for the synthetic periodic workloads we use in
> +            # LISA (where all threads wake up at the same time) but is probably
> +            # no good for real workloads.
> +            active_time = max(cpu_active_time[c] for c in cpus)
> +            active_power = node.active_states[freq].power * active_time
> +
> +            _idle_power = max(node.idle_states[idle_states[c]] for c in cpus)
> +            idle_power = _idle_power * (1 - active_time)
> +
> +            if combine:
> 
> Hmm, I actually thought I had removed this parameter completely, it seemed useful at first but in the end I never ended up using it, would that be OK with you?

You mean if it's ok to remove it?
I personally like the idea to have a split version... that's why I was
just proposing to use a namedtuple to get rid of the parameter but
still keep the active and idle contributions available.

> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ARM-software/lisa/pull/263#discussion_r102735588

-- 
#include <best/regards.h>

Patrick Bellasi
